src/helper.cc

```diff
@@ -851,15 +853,18 @@ helper::~helper()
 {
     /* note, don't free id_name, it probably points to static memory */
 
-    // TODO: if the queue is not empty it will leak Helper::Request's
-    if (!queue.empty())
-        debugs(84, DBG_CRITICAL, "WARNING: freeing " << id_name << " helper with " << stats.queue_size << " requests queued");
+    // An non-empty queue would leak Helper::Xaction objects, stalling any
+    // pending (and even future collapsed) transactions. A live server will keep
+    // us, its parent, alive due to reference counting. The idle=1 minimum means
+    // that either there has to be at least one server alive, or the last
+    // attempt to start a server has failed. If such an attempt fails,
+    // handleKilledServer() is called to drain the queue.
```

Comment:
Capitalizing on the smaller change request logic, we can also say that it does not really matter what exactly leads to the "no servers" condition. What matters is that when we are being destroyed, there must be no queued requests left. This allows us to simplify further _and_ document why we are asserting instead of calling dropQueued() here:

```suggestion
    // pending (and even future collapsed) transactions. To avoid stalling
    // transactions, we must dropQueued(). We ought to do that when we
    // discover that no progress is possible rather than here, so that we do
    // no stall transactions if reference counting keeps us alive longer.
```

Reply:
I used this version (with a minor modification to avoid repetition).